### ISSUE TITLE
ListView scroll position jumps when items change

### ISSUE THREAD

**Issue by hettiger** [NativeScript/NativeScript#5588]
### Did you verify this is a real problem by searching the [NativeScript Forum](http://forum.nativescript.org) and the [other open issues in this repo](https://github.com/NativeScript/nativescript/issues)?

Yes

### Tell us about the problem

On iOS the ListView scroll position jumps when items change. This seems to only happen with variable item heights. Android handles this perfectly fine.

I think this is a pretty serious bug because it makes it pretty much impossible to use load more items in real world apps. Unfortunately RadListView is not an alternative because it has some serous bugs too. (Which made me try to implement everything with ListView which seems to be much more stable)

### Which platform(s) does your issue occur on?
iOS

### Please provide the following version numbers that your issue occurs with:

The issue is reproducible in the official {NS} playgrounds:

```json
"nativescript": {
    "id": "org.nativescript.blankng",
    "tns-android": {
        "version": "3.4.1"
    },
    "tns-ios": {
        "version": "3.4.1"
    },
    "playground": {
        "id": "1606970604.1480346322"
    }
},
"dependencies": {
    "@angular/animations": "~5.2.0",
    "@angular/common": "~5.2.0",
    "@angular/compiler": "~5.2.0",
    "@angular/core": "~5.2.0",
    "@angular/forms": "~5.2.0",
    "@angular/http": "~5.2.0",
    "@angular/platform-browser": "~5.2.0",
    "@angular/platform-browser-dynamic": "5.2.0",
    "@angular/router": "~5.2.0",
    "nativescript-accelerometer": "^2.0.1",
    "nativescript-angular": "~5.2.0",
    "nativescript-geolocation": "^4.2.3",
    "nativescript-ui-chart": "^3.5.0",
    "nativescript-ui-listview": "^3.5.1",
    "nativescript-ui-sidedrawer": "^3.5.1",
    "nativescript-ui-calendar": "^3.5.0",
    "nativescript-ui-autocomplete": "^3.6.0",
    "nativescript-ui-dataform": "^3.5.0",
    "nativescript-ui-gauge": "^3.5.0",
    "nativescript-theme-core": "~1.0.4",
    "kinvey-nativescript-sdk": "^3.10.0",
    "nativescript-camera": "^4.0.0",
    "nativescript-fresco": "^3.0.6",
    "nativescript-intl": "^3.0.0",
    "nativescript-iqkeyboardmanager": "^1.2.0",
    "nativescript-social-share": "^1.5.0",
    "reflect-metadata": "~0.1.12",
    "rxjs": "~5.5.6",
    "tns-core-modules": "~3.4.1",
    "zone.js": "~0.8.20"
},
"devDependencies": {
    "babel-traverse": "6.25.0",
    "babel-types": "6.25.0",
    "babylon": "6.17.4",
    "lazy": "1.0.11",
    "nativescript-dev-typescript": "~0.6.0",
    "typescript": "~2.4.2"
}
```

### Please tell us how to recreate the issue in as much detail as possible.

https://play.nativescript.org/?template=play-ng&id=YWJxIM&v=8

Simply scroll down until the load more items loading indicator appears and observe the scroll position closely. Because of the variable item heights the scroll jumps can appear smaller or bigger from load to load.

**Comment by hettiger** [comment]
I found a solution:

```ts
onListViewLoaded(args): void {
    this.listView = args.object;

    if (isIOS) {
        /**
         * Enforce UITableViewDelegateImpl
         *
         * Make sure UITableViewDelegateImpl is being used because that's the
         * tns-core-modules/ui/list-view delegate that is capable of correctly
         * measuring row heights.
         *
         * @see https://github.com/NativeScript/NativeScript/blob/1931669f79d4177eb2d7c90fe7573a1f18751eb7/tns-core-modules/ui/list-view/list-view.ios.ts#L300
         */
        this.listView.rowHeight = -1;

        /**
         * Enforce precise row height's
         *
         * Set the estimatedRowHeight to 0 in order to make sure that rows that
         * are currently off the screen will be measured precisely instead of just
         * using the estimated row height until they get scrolled into view.
         * This is very important in order to maintain the correct scroll position
         * when loading more items.
         */
        this.listView.ios.estimatedRowHeight = 0;
    }
}
```

Are my comments correct or why exactly is this fixing it? I guess this is reducing performance on long lists but it seems to be a requirement for load on demand + variable heights. Can we add an optional "precise mode" to the core?

**Comment by hettiger** [comment]
Short follow up: I've now completed the switch from RadListView to ListView in my entire app. Everything's much more stable now and I have no performance issues whatsoever using the above fix with very long lists.

**Comment by NickIliev** [comment]
possibly related to [this fix](https://github.com/NativeScript/NativeScript/pull/5693) - @hettiger let us know if the issue is resolved with tns-core-modules@next (if you have sample project that you can easily update)

**Comment by hettiger** [comment]
@NickIliev Unfortunately it is not related. But my workaround is still performing great. Even with thousands of items in the list. I have no performance issues whatsoever.

Heres my test setup:

```json
    "nativescript": {
        "id": "org.nativescript.blankng",
        "tns-android": {
            "version": "4.0.1"
        },
        "tns-ios": {
            "version": "4.0.1"
        }
    },
    "dependencies": {
        "@angular/animations": "~5.2.0",
        "@angular/common": "~5.2.0",
        "@angular/compiler": "~5.2.0",
        "@angular/core": "~5.2.0",
        "@angular/forms": "~5.2.0",
        "@angular/http": "~5.2.0",
        "@angular/platform-browser": "~5.2.0",
        "@angular/platform-browser-dynamic": "5.2.0",
        "@angular/router": "~5.2.0",
        "kinvey-nativescript-sdk": "^3.10.2",
        "nativescript-accelerometer": "^2.0.1",
        "nativescript-angular": "~5.2.0",
        "nativescript-background-http": "^3.2.2",
        "nativescript-camera": "^4.0.2",
        "nativescript-fresco": "^3.0.6",
        "nativescript-geolocation": "^4.2.3",
        "nativescript-intl": "^3.0.0",
        "nativescript-iqkeyboardmanager": "^1.3.0",
        "nativescript-social-share": "^1.5.0",
        "nativescript-theme-core": "~1.0.4",
        "nativescript-ui-autocomplete": "^3.6.0",
        "nativescript-ui-calendar": "^3.5.2",
        "nativescript-ui-chart": "^3.5.0",
        "nativescript-ui-dataform": "^3.5.2",
        "nativescript-ui-gauge": "^3.5.0",
        "nativescript-ui-listview": "^3.5.4",
        "nativescript-ui-sidedrawer": "^4.0.0",
        "reflect-metadata": "~0.1.12",
        "rxjs": "~5.5.6",
        "tns-core-modules": "^4.0.0-2018-04-26-02",
        "zone.js": "~0.8.20"
    },
    "devDependencies": {
        "babel-traverse": "6.25.0",
        "babel-types": "6.25.0",
        "babylon": "6.17.4",
        "lazy": "1.0.11",
        "nativescript-dev-typescript": "~0.6.0",
        "typescript": "~2.4.2"
    }
```

(I downloaded the [Playground](https://play.nativescript.org/?template=play-ng&id=YWJxIM&v=9) and ran `tns-core-modules@next` before building & running the app. Can still reproduce the issue.)

**Comment by NickIliev** [comment]
@hettiger after some internal investigation it seems that thе issue is related to how table cells are sized on iOS (see here for [detailed explanation](https://developer.apple.com/library/content/documentation/UserExperience/Conceptual/AutolayoutPG/WorkingwithSelf-SizingTableViewCells.html)). We could hardcore a default value for [`iosEstimatedRowHeight`](https://docs.nativescript.org/api-reference/classes/_ui_list_view_.listview#iosestimatedrowheight) (which is the NativeScript property equivalent of iOS [`estimatedRowHeight`](https://developer.apple.com/documentation/uikit/uitableview/1614925-estimatedrowheight) ) but the overall conclusion is that there should be no default value and the control should be in the hand of the developers

So to resolve your case, you could use your workaround but directly by using the NativeScript's property (which is also more performance optimized)

For example:
```
onListViewLoaded(args): void {
    this.listView = args.object;

    if (isIOS) {
        this.listView.iosEstimatedRowHeight = 0; // or your own preffered value
    }
}
```

[Playground demo showing the above approach.](https://play.nativescript.org/?template=play-ng&id=YWJxIM&v=14)

**Comment by hettiger** [comment]
@NickIliev Thank you for the clarification. However the documentation needs to improve for sure. It took me quite some time to find this out myself. Theres no mention of the `iosEstimatedRowHeight` or `rowHeight` properties at the [ListView Documentation](https://docs.nativescript.org/angular/code-samples/ui/listview). Also the `nsTemplateKey` is lacking documentation. I found this fortunately by reading in the `nativescript-angular` source code. Do you want me to draft a PR as a starting point?

**Comment by NickIliev** [comment]
@hettiger of course, all PRs are more than welcomed :) [This is the DOCS repository](**url**) - keep in mind that the `nsTemplateKey` is described in the Angular falvor of the documentation in [this article section](https://docs.nativescript.org/angular/ui/list-view#using-multiple-item-templates) (there are HTML and TS snippets) and also there are additional examples [here](https://docs.nativescript.org/angular/code-samples/common-screens/lists/two-lines-grouped), [here](https://docs.nativescript.org/angular/code-samples/common-screens/lists/single-line-grouped), [here](https://docs.nativescript.org/angular/code-samples/common-screens/lists/multi-line-grouped) and [here](https://docs.nativescript.org/angular/code-samples/common-screens/lists/multi-line-big). However, we do lack an example for `iosEstimatedRowHeight`  property!

**Comment by hettiger** [comment]
@NickIliev Feeling really blind now. Have you updated the docs lately? I will draft a PR as soon as possible. Thank you.

**Comment by hettiger** [comment]
@NickIliev I've just tested your playground to see if it is really enough to omit `this.listView.rowHeight = -1;` and I cannot confirm this. Also you have set the `iosEstimatedRowHeight` to `100` which will not work either. Anyways it is not your fault. On the first load on demand attempts that I've made it didn't jump either. That's because the estimated row height that you've set seems to be working really well with the content. If you make more attempts and maybe refresh the list a few times you will see it is indeed jumping sometimes. A way to better reproduce is to introduce more variation when it comes to the item heights. I've made some changes to the playground that make it easier for you to reproduce:

https://play.nativescript.org/?template=play-ng&id=YWJxIM&v=17

I think this `rowHeight = -1;` and `iosEstimatedRowHeight = 0;` is the only way to go right now. Anyways it feels like an ugly fix and I'm worried it could probably break in future versions. Do you really want me to document it like that or isn't something like a precise mode a more viable approach?

**Comment by NickIliev** [comment]
@hettiger you are right and in my test app, the 100 value was leftover from different test cases.
Setting the `iosEstimatedRowHeight` to 0 should work
```
    if (isIOS) {
        this.listView.iosEstimatedRowHeight = 0; 
    }
```

**Comment by hettiger** [comment]
@NickIliev Alright awesome. Now I can confirm it is working. Docs PR is coming. Thank you 👍

**Comment by FranciZ** [comment]
Would it be possible to just hardcode `iosEstimatedRowHeight = 0` as a default for ListView? This issue caused me so much headache that I really think other people should be speared debugging this by just setting the default value to 0. Or is there some reason not to do so?
